external/mlir-hal/lib/Dialect/MHAL/Transforms/Prefill.cpp

```diff
@@ -0,0 +1,140 @@
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/MHAL/IR/MHAL.h"
+#include "mlir/Dialect/MHAL/Transforms/Passes.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/Location.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/RegionUtils.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace mlir {
+namespace mhal {
+#define GEN_PASS_DEF_MHALPREFILLPASS
+#include "mlir/Dialect/MHAL/Transforms/Passes.h.inc"
+} // namespace mhal
+} // namespace mlir
+
+#define DEBUG_TYPE "mhal-prefill"
+
+using namespace mlir;
+namespace {
+class MHALPrefillPass
+    : public mhal::impl::MHALPrefillPassBase<MHALPrefillPass> {
+
+  Value createConstantOp(OpBuilder &builder, Type elementType,
+                         uint32_t initValue) {
+    Value constantOp;
+    auto loc = builder.getUnknownLoc();
+    if (llvm::isa<FloatType>(elementType)) {
+      auto semantics = static_cast<APFloat::Semantics>(-1);
+      if (elementType.isF32()) {
+        semantics = APFloat::S_IEEEsingle;
+      } else if (elementType.isF16()) {
+        semantics = APFloat::S_IEEEhalf;
+      } else if (elementType.isBF16()) {
+        semantics = APFloat::S_BFloat;
+      } else if (elementType.isFloat8E4M3FNUZ()) {
+        semantics = APFloat::S_Float8E4M3FNUZ;
+      } else if (elementType.isFloat8E5M2FNUZ()) {
+        semantics = APFloat::S_Float8E5M2FNUZ;
+      } else {
+        llvm_unreachable("Unexpected float semantics");
+      }
+
+      APFloat apValue(*(reinterpret_cast<float *>(&initValue)));
+      bool lostInfo = false;
+      apValue.convert(APFloat::EnumToSemantics(semantics),
+                      APFloat::rmNearestTiesToEven, &lostInfo);
+
+      auto constValue = builder.getFloatAttr(elementType, apValue);
+      constantOp =
+          builder.create<arith::ConstantOp>(loc, elementType, constValue);
+    } else {
+      APInt apValue(elementType.getIntOrFloatBitWidth(), initValue, true);
+      auto constValue = builder.getIntegerAttr(elementType, apValue);
+      constantOp =
+          builder.create<arith::ConstantOp>(loc, elementType, constValue);
+    }
+    return constantOp;
+  }
+
+  void insertPrefillOps(OpBuilder &builder, gpu::LaunchFuncOp &launchOp) {
+    auto func = llvm::cast<func::FuncOp>(launchOp->getParentOp());
```

Comment:
Shouldn't need `llvm::` here. It should just be `cast` by way of a `using namespace mlir;`, which reexports `llvm` in some cases

Reply:
It is good to know. Thanks!